lib/spack/spack/spec.py

```diff
@@ -2617,10 +2679,11 @@ def normalize(self, force=False, tests=False, user_spec_deps=None):
                 else:
                     all_spec_deps[name].constrain(spec)
 
-        # Initialize index of virtual dependency providers if
-        # concretize didn't pass us one already
-        provider_index = spack.provider_index.ProviderIndex(
-            [s for s in all_spec_deps.values()], restrict=True)
+        # FIXME: Revisit this part
+        bindings = getattr(self, '_explicit_providers', {})
+        provider_index = spack.provider_index.IndexWithBindings(
+            [s for s in all_spec_deps.values()], bindings
+        )
```

Comment:
I think the bindings should be parsed into attributes *on the DependencySpec* of each node.  So, if I have:

```
hdf5 ^mpi=mpich
```

Then I should get a spec that looks like this
```
hdf5
 |
 | this edge should have an mpi attribute
 | (i.e. its DependencySpec should say it provides mpi)
 v
mpich
```

That `mpi` attribute should also get preserved in the final `spec.yaml`.  I *think* we only need to provide names here, not full virtual specs, as we don't really know or care what version of the virtual is actually being "used" by the dependent (the concretizer just needs to ensure that a suitable mpich configuration is resolved)

Reply:
@tgamblin Do we want these properties on edges to signal a user requested binding or do we want to add an attribute to edges to represent that one or more virtual request are satisfied by the edge? What I mean is that after concretizing:
```
hdf5 ^mpich
```
I expect to have an edge with the same attribute on the concretized spec:
```
hdf5
 |
 | this edge should have an mpi attribute
 | (i.e. its DependencySpec should say it provides mpi)
 v
mpich
```
and the attribute is reflected in the hash of the DAG. Am I correct on this?

Reply:
As of b91f20f:
```console
$ spack spec -y netlib-scalapack ^lapack,blas=openblas ^mpi=intel-parallel-studio@cluster
spec:
- netlib-scalapack:
    version: 2.1.0
    arch:
      platform: linux
      platform_os: ubuntu18.04
    ...
    dependencies:
      cmake:
        hash: svn7w7iny7wdbuziefk4pk6kqkmkrozr
        type:
        - build
        virtuals: []
      intel-parallel-studio:
        hash: 3f3kxikxcaf7y526aj4lwujf3zxpg5cx
        type:
        - build
        - link
        virtuals:
        - mpi
      openblas:
        hash: wqwrkprfr27st77niwdys6ndsxx5bvsj
        type:
        - build
        - link
        virtuals:
        - blas
        - lapack
    hash: qhvpvatgqtmtqytedgc6v76nk34l4xtb
    build_hash: bfjx7xjiyroabgjfcobxfcza4pezf34q
- cmake:
  ...
```

Reply:
Nice!  Can we call it `provides:`?

Reply:
Done in 61cc729